workflow/controller/controller.go

```diff
@@ -569,27 +594,33 @@ func (wfc *WorkflowController) processNextPodCleanupItem(ctx context.Context) bo
 	return true
 }
 
-func (wfc *WorkflowController) getPod(namespace string, podName string) (*apiv1.Pod, error) {
-	obj, exists, err := wfc.podInformer.GetStore().GetByKey(namespace + "/" + podName)
-	if err != nil {
-		return nil, err
-	}
-	if !exists {
-		return nil, nil
-	}
-	pod, ok := obj.(*apiv1.Pod)
-	if !ok {
-		return nil, fmt.Errorf("object is not a pod")
+func createFinalizerPodStatusRemovalPatchIfExists(pod *apiv1.Pod) []PatchOperation {
+	var patch []PatchOperation
+	i := slices.Index(pod.Finalizers, common.FinalizerPodStatus)
+	if i >= 0 {
+		patch = append(patch, PatchOperation{
+			Operation: "remove",
+			Path:      fmt.Sprintf("/metadata/finalizers/%d", i),
+		})
 	}
-	return pod, nil
+	return patch
 }
 
-func (wfc *WorkflowController) signalContainers(namespace string, podName string, sig syscall.Signal) (time.Duration, error) {
-	pod, err := wfc.getPod(namespace, podName)
-	if pod == nil || err != nil {
-		return 0, err
+func applyPatchIfExists(ctx context.Context, pods typedv1.PodInterface, podName string, patch []PatchOperation) error {
+	if len(patch) == 0 {
+		log.Debug("not patching pod")
+		return nil
 	}
+	data, err := json.Marshal(patch)
+	if err != nil {
+		return fmt.Errorf("failed to marshal patch: %w", err)
+	}
+	log.WithField("data", string(data)).Debug("patching pod")
+	_, err = pods.Patch(ctx, podName, types.JSONPatchType, data, metav1.PatchOptions{})
+	return err
+}
```

Comment:
Nit: should this be named something like `applyPatches()` since there are more than one? (and since it's an array slice, I don't even feel we need "IfExists" in the name, since any array slice can be of 0 length, although okay if you do want that)

Reply:
I have addressed this in the commit in the following comment.
https://github.com/argoproj/argo-workflows/pull/12413#discussion_r1446980034